Fix deleted items auto expiration for users with no quota

VicDeo opened the pull request:
## Description
Fix regression introduced a long time ago in https://github.com/owncloud/core/commit/a3999036f720e73727dac07a337578f3e25f92aa#diff-0dc914768208fc24b2aa2f2a369bb684
It is based on a wrong assumption that FS free space could be negative. 

## Related Issue
https://github.com/owncloud/core/issues/27412
obsoletes https://github.com/owncloud/core/pull/29951
## Motivation and Context
Trashbin auto clean up doesn't work when a user has no quota

## How Has This Been Tested?
1. Set `trashbin_retention_obligation` to `auto`
2. Create a user with no quota
3. Upload some files and delete uploaded files for that user
4. Keep uploading until all partition for `datadir` is occupied

### Expected 
Deleted files are purged to save space

### Actual
They are not


## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

To document: 
New config.php `'trashbin_purge_limit'` that was just a constant before.
Introduced to fine tune purging in cases like https://github.com/owncloud/core/issues/2936

VicDeo committed:
Refactor Trashbin::calculateFreeSpace to be testable

PVince81 commented:
@VicDeo that's fantastic ! Can you quickly explain what approach you used to fix it ?

The commit is called "Refactor Trashbin::calculateFreeSpace to be testable" which doesn't imply a fix and I can't seem to quickly find the actual location in the code where the fix is happening.

Suggestion: have one commit for the testable refactor and another one with the actual fix.

VicDeo commented:
@PVince81 
I just reverted the changes from the previous test and moved the method into another non-static class

As soon as it will reintroduce purging of deleted files described in  https://github.com/owncloud/core/issues/2936 
`trashbin_purge_limit` could be set to 30%, 20%, 10%, etc in case there are no users with hard quota

PVince81 reviewed (approved):
Makes sense, thanks 👍

PVince81 commented:
Please backport

VicDeo commented:
Stable10: https://github.com/owncloud/core/pull/30163
